src/OrchardCore.Modules/OrchardCore.Microsoft.Authentication/Configuration/MicrosoftAccountSettingsConfiguration.cs

```diff
@@ -0,0 +1,51 @@
+using System;
+using Microsoft.AspNetCore.DataProtection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using OrchardCore.Microsoft.Authentication.Services;
+using OrchardCore.Microsoft.Authentication.Settings;
+
+namespace OrchardCore.Microsoft.Authentication.Configuration;
+
+public class MicrosoftAccountSettingsConfiguration : IConfigureOptions<MicrosoftAccountSettings>
+{
+    private readonly IMicrosoftAccountService _microsoftAccountService;
+    private readonly IDataProtectionProvider _dataProtectionProvider;
+    private readonly ILogger _logger;
+
+    public MicrosoftAccountSettingsConfiguration(
+        IMicrosoftAccountService microsoftAccountService,
+        IDataProtectionProvider dataProtectionProvider,
+        ILogger<MicrosoftAccountSettingsConfiguration> logger)
+    {
+        _microsoftAccountService = microsoftAccountService;
+        _dataProtectionProvider = dataProtectionProvider;
+        _logger = logger;
+    }
+
+    public void Configure(MicrosoftAccountSettings options)
+    {
+        var settings = _microsoftAccountService
+            .GetSettingsAsync()
+            .GetAwaiter()
+            .GetResult();
+
+        options.AppId = settings.AppId;
+        options.CallbackPath = settings.CallbackPath;
+        options.SaveTokens = settings.SaveTokens;
+
+        if (!String.IsNullOrWhiteSpace(settings.AppSecret))
+        {
+            try
+            {
+                var protector = _dataProtectionProvider.CreateProtector(nameof(MicrosoftAccountSettingsConfiguration));
+
+                options.AppSecret = protector.Unprotect(settings.AppSecret);
```

Comment:
We can't expect people to encrypt the app secret with the current data protection key when they manually configure an external configuration like this, especially that these keys are auto-rotated periodically [and shouldn't be used to encrypt for storage](https://andrewlock.net/an-introduction-to-the-data-protection-system-in-asp-net-core/#is-there-anything-i-shouldn-t-use-data-protection-for-). So, I'd just use the direct value here.

Same for all the others.

Reply:
Seems we already using a data protection here

https://github.com/OrchardCMS/OrchardCore/blob/2db2808a13b304c62b9063d142201440885e9bb8/src/OrchardCore.Modules/OrchardCore.Microsoft.Authentication/Configuration/MicrosoftAccountOptionsConfiguration.cs#L73

Reply:
`IOptions` can be anything. Usually appsettings.json, yes, but frequently also environment variables. In any case, it's not a configuration source that the Orchard app will write to (and thus be able to use `IDataProtectionProvider` to encrypt values), so, it shouldn't use `IDataProtectionProvider` to decrypt either.

So, to be clear:

- En/decrypting values with `IDataProtectionProvider` stored in site settings or otherwise in the DB is OK (OK-ish).
- Decrypting values with `IDataProtectionProvider` coming from configuration providers should never happen.

Reply:
That's what I realize from your previous comment, so will resolve the conversation while it's already done ;)

Reply:
Wait, the referenced line only runs for site settings? In that case, I'm sorry about the confusion, and nothing to change.